<a href="https://colab.research.google.com/github/simmy4946/NEW-REPOSITS/blob/main/Car_price_prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Import Libraries:
import pandas as pd

1.Calculate aggregate statistics such as mean, median, and standard deviation of car prices within each group
2.What is the average price of cars for each make and model?
3.Is there a relationship between the year of the car and its price?
4.How does the condition of the car affect its price?
5.What is the correlation between price and numerical features (mileage, year)?
6.What is the average price of cars for each make and model?
7.Which brands and models have the highest and lowest average prices?
8.How have car prices changed over the years?
9.Is there a noticeable trend in pricing based on the manufacturing year?
10.Is there a correlation between the price of the car and its mileage?
11.Do cars with higher mileage tend to have lower prices?
12.How does the condition of the car (Excellent, Good, Fair) affect its price?
13.Are there significant differences in prices based on the condition of the car?
14.Can you compare the average prices of cars from different manufacturers?
15.How do prices vary between different car models within the same manufacturer?
16.Calculate summary statistics like mean, median, and standard deviation of car prices to describe the central tendency and variability of the data.

In [ ]:
# Import Data:
df = pd.read_csv("/content/CarPricesPrediction.csv")

In [ ]:
# see the dataset:
df

Unnamed: 0       Make      Model  Year  Mileage  Condition     Price
0             0       Ford  Silverado  2022    18107  Excellent  19094.75
1             1     Toyota  Silverado  2014    13578  Excellent  27321.10
2             2  Chevrolet      Civic  2016    46054       Good  23697.30
3             3       Ford      Civic  2022    34981  Excellent  18251.05
4             4  Chevrolet      Civic  2019    63565  Excellent  19821.85
..          ...        ...        ...   ...      ...        ...       ...
995         995     Nissan      Camry  2010   149032  Excellent  24548.50
996         996  Chevrolet      F-150  2014    20608  Excellent  26969.70
997         997       Ford     Altima  2016   109851       Good  20507.55
998         998     Toyota  Silverado  2010    11704       Good  31414.90
999         999     Nissan  Silverado  2017   128390  Excellent  18580.60

[1000 rows x 7 columns]

In [ ]:
# see the top 5 heads of data:
df.head(5)

Unnamed: 0       Make      Model  Year  Mileage  Condition     Price
0           0       Ford  Silverado  2022    18107  Excellent  19094.75
1           1     Toyota  Silverado  2014    13578  Excellent  27321.10
2           2  Chevrolet      Civic  2016    46054       Good  23697.30
3           3       Ford      Civic  2022    34981  Excellent  18251.05
4           4  Chevrolet      Civic  2019    63565  Excellent  19821.85

In [ ]:
# see the shape of data :
df.shape

(1000, 7)

In [ ]:
# see the columns:
df.columns

Index(['Unnamed: 0', 'Make', 'Model', 'Year', 'Mileage', 'Condition', 'Price'], dtype='object')

In [ ]:
# see the size of data:
df.size

7000

In [ ]:
# get the information of the data:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  1000 non-null   int64  
 1   Make        1000 non-null   object 
 2   Model       1000 non-null   object 
 3   Year        1000 non-null   int64  
 4   Mileage     1000 non-null   int64  
 5   Condition   1000 non-null   object 
 6   Price       1000 non-null   float64
dtypes: float64(1), int64(3), object(3)
memory usage: 54.8+ KB


In [ ]:
# check the null values of the data:
df.isnull().sum() # there is no null values in the dataset

Unnamed: 0    0
Make          0
Model         0
Year          0
Mileage       0
Condition     0
Price         0
dtype: int64

In [ ]:
# see the statistical summary of the data:
df.describe


<bound method NDFrame.describe of      Unnamed: 0       Make      Model  Year  Mileage  Condition     Price
0             0       Ford  Silverado  2022    18107  Excellent  19094.75
1             1     Toyota  Silverado  2014    13578  Excellent  27321.10
2             2  Chevrolet      Civic  2016    46054       Good  23697.30
3             3       Ford      Civic  2022    34981  Excellent  18251.05
4             4  Chevrolet      Civic  2019    63565  Excellent  19821.85
..          ...        ...        ...   ...      ...        ...       ...
995         995     Nissan      Camry  2010   149032  Excellent  24548.50
996         996  Chevrolet      F-150  2014    20608  Excellent  26969.70
997         997       Ford     Altima  2016   109851       Good  20507.55
998         998     Toyota  Silverado  2010    11704       Good  31414.90
999         999     Nissan  Silverado  2017   128390  Excellent  18580.60

[1000 rows x 7 columns]>

In [ ]:
# drop the unneccessary column of the data:
df.drop(columns=['Unnamed: 0'], inplace=True)
#df.drop(columns=['Unnamed: 0'], axis=1)

df.head(2)

Make      Model  Year  Mileage  Condition     Price
0    Ford  Silverado  2022    18107  Excellent  19094.75
1  Toyota  Silverado  2014    13578  Excellent  27321.10

In [ ]:
# see the datatypes :
df.dtypes

Make          object
Model         object
Year           int64
Mileage        int64
Condition     object
Price        float64
dtype: object

In [ ]:
# check the duplicates:
df_no_duplicates = df.drop_duplicates() # there is no duplicates in this dataset
df.shape

(1000, 6)

In [ ]:
# frequency (count)of each category (make) of car:
df['Make'].value_counts()

Chevrolet    209
Toyota       205
Ford         199
Honda        199
Nissan       188
Name: Make, dtype: int64

In [ ]:
# check the count of each category of condition column of data:
df['Condition'].value_counts()

Excellent    595
Good         292
Fair         113
Name: Condition, dtype: int64

In [ ]:
# set both of them together:
df[['Make','Condition']].value_counts().reset_index()

Make  Condition    0
0      Toyota  Excellent  126
1   Chevrolet  Excellent  120
2        Ford  Excellent  120
3       Honda  Excellent  118
4      Nissan  Excellent  111
5   Chevrolet       Good   69
6        Ford       Good   57
7      Nissan       Good   57
8       Honda       Good   55
9      Toyota       Good   54
10      Honda       Fair   26
11     Toyota       Fair   25
12       Ford       Fair   22
13  Chevrolet       Fair   20
14     Nissan       Fair   20

**1. How many manufacturers for cars making and their names?**

In [ ]:
unique_makes = df['Make'].unique()
unique_makes

array(['Ford', 'Toyota', 'Chevrolet', 'Nissan', 'Honda'], dtype=object)

There are 5 unique manufacturer in this data above

**2.how many unique models for cars in data and their names?**

In [ ]:
unique_model = df['Model'].unique()
unique_model

array(['Silverado', 'Civic', 'Altima', 'Camry', 'F-150'], dtype=object)

there are 5 unique models in this data above

In [ ]:
df['Make'].nunique()

5

In [ ]:
df['Model'].nunique()

5

**3.Which is the maximum and minimum value of manufaturing(Make)of cars?**

In [ ]:
df['Make'].max()

'Toyota'

maximum value of make column or manufacturers is "Toyota"

In [ ]:
df['Make'].min()

'Chevrolet'

minimum value of make column or manufacturers is "Chevrolet"



maximum value of make column or manufacturers is "Toyota"

**4.Which is the maximum and minimum value of model of cars or which model is highest and lowest in frequency ?**

In [ ]:
df['Model'].max()

'Silverado'

Maximum number of cars of model 'Silerado'.

In [ ]:
df['Model'].min()

'Altima'

Minimum number of cars of model 'Altima'

**4.What is the average price,median price and standard deviation or variance of cars for each make and model?**

In [ ]:
agg= df.groupby(['Make','Model'])['Price'].agg(['mean','median','std'])
agg

mean     median          std
Make      Model                                          
Chevrolet Altima     21407.802000  22210.600  4379.431219
          Camry      21912.704878  21920.650  4181.601382
          Civic      21721.005263  21190.575  4727.547017
          F-150      22464.695652  22508.200  4364.854807
          Silverado  22396.186765  22693.375  4346.153394
Ford      Altima     23576.052632  24893.325  4012.856885
          Camry      21660.350000  20737.625  4271.645547
          Civic      22853.082895  22698.075  4196.766261
          F-150      21627.950000  22045.900  4059.090390
          Silverado  22250.963889  22588.025  3704.510871
Honda     Altima     21520.601250  22052.800  4527.914662
          Camry      24099.205000  24339.375  3481.001669
          Civic      21779.972973  20856.150  4522.788369
          F-150      22866.732258  23033.450  3922.139683
          Silverado  21291.349020  21318.200  3458.981578
Nissan    Altima     21493.059783  21090.350  4218.806553
          Camry      23144.928750  23264.700  4099.125278
          Civic      22029.708065  21542.100  4585.320077
          F-150      24045.532857  23683.600  3675.195838
          Silverado  22429.386111  23536.700  4411.034937
Toyota    Altima     21547.499038  21927.475  4197.593157
          Camry      21762.431395  22276.800  4651.103640
          Civic      21735.738158  21531.500  4069.313471
          F-150      22429.305405  23427.550  4553.521352
          Silverado  22105.582857  22087.050  4770.531338

Average, median price and sd for checking the variation of cars for each 'make' and 'model'

**5.What is the average price and other stats of cars for the 'conditions' in the data?**

In [ ]:
agg= df.groupby(['Condition'])['Price'].agg(['mean','median','std'])
agg

mean     median          std
Condition                                      
Excellent  22150.842017  22130.450  4198.703137
Fair       22357.820796  22544.700  4480.722085
Good       22222.674315  22331.225  4259.192925

from above, we do groupby 'condition' of cars with price to derived their mean, median and sd.

**6.What is the average mileage and median & sd of cars for each model?**

In [ ]:
agg= df.groupby(['Model'])['Mileage'].agg(['mean','median','std'])
agg

mean   median           std
Model                                         
Altima     80332.526549  78038.0  39430.031785
Camry      81730.471698  85209.5  41783.363346
Civic      73346.807692  70872.0  38502.821121
F-150      74659.223404  72131.0  38348.899491
Silverado  82968.036458  84643.5  40285.105760

from above, we do groupby of model with mileage to find their mean, median and sd.

**7.What is the average price of cars for each make and model?**

In [ ]:
avg_price_cars= df.groupby(['Make','Model'])['Price'].agg(['mean'])
avg_price_cars.sort_values(by= 'mean')

mean
Make      Model                  
Honda     Silverado  21291.349020
Chevrolet Altima     21407.802000
Nissan    Altima     21493.059783
Honda     Altima     21520.601250
Toyota    Altima     21547.499038
Ford      F-150      21627.950000
          Camry      21660.350000
Chevrolet Civic      21721.005263
Toyota    Civic      21735.738158
          Camry      21762.431395
Honda     Civic      21779.972973
Chevrolet Camry      21912.704878
Nissan    Civic      22029.708065
Toyota    Silverado  22105.582857
Ford      Silverado  22250.963889
Chevrolet Silverado  22396.186765
Toyota    F-150      22429.305405
Nissan    Silverado  22429.386111
Chevrolet F-150      22464.695652
Ford      Civic      22853.082895
Honda     F-150      22866.732258
Nissan    Camry      23144.928750
Ford      Altima     23576.052632
Nissan    F-150      24045.532857
Honda     Camry      24099.205000

average price of cars for each 'make' and 'model', highest mean of make is "Honda" with model "Camry" and lowest mean of 'make' is 'Honda' and 'model' 'silverado'.


**8.What is the average mileage of cars for each make and model?**

In [ ]:
avg_mileage_cars= df.groupby(['Make','Model'])['Mileage'].agg(['mean'])
avg_mileage_cars.sort_values(by ='mean')

mean
Make      Model                   
Ford      Civic       63992.078947
Nissan    F-150       65947.285714
Ford      Altima      69006.421053
Chevrolet F-150       70272.347826
Nissan    Civic       71019.838710
          Camry       71102.525000
Chevrolet Civic       72949.368421
Nissan    Altima      74922.282609
Honda     F-150       75569.935484
Ford      Camry       77210.583333
Toyota    Civic       78444.289474
Honda     Camry       78516.700000
Nissan    Silverado   79191.055556
Chevrolet Camry       79795.902439
Honda     Civic       80076.918919
Ford      F-150       80262.333333
          Silverado   81092.722222
Toyota    Altima      81358.980769
          F-150       81685.189189
Honda     Silverado   82017.450980
Chevrolet Silverado   82665.323529
Honda     Altima      83089.075000
Toyota    Silverado   90461.028571
Chevrolet Altima      90645.040000
Toyota    Camry      101496.534884

average mileage of cars for each 'make' and 'model', highest mean of make is "Toyota" with model "Camry" and lowest mean of 'make' is 'Ford' and 'model' 'Civic'.


**9.Which brands and models have the highest and lowest average prices? ***

In [ ]:
avg_price_cars_mod= df.groupby('Model')['Price'].mean().sort_values()
avg_price_cars_mod

Model
Altima       21841.835619
Civic        22025.018407
Silverado    22028.735156
Camry        22470.130189
F-150        22644.748404
Name: Price, dtype: float64

In [ ]:
avg_price_cars_mod.idxmax()

'F-150'

In [ ]:
avg_price_cars_mod.idxmin()

'Altima'

The Highest average price of model is F-150.
The Lowest average price of model is Altima.


In [ ]:
avg_price_cars_mak= df.groupby('Make')['Price'].mean().sort_values()
avg_price_cars_mak

Make
Toyota       21881.913415
Chevrolet    21957.203589
Honda        22238.084171
Ford         22354.415578
Nissan       22587.502128
Name: Price, dtype: float64

In [ ]:
avg_price_cars_mak.idxmin()

'Toyota'

In [ ]:
avg_price_cars_mak.idxmax()

'Nissan'

The Highest average price of make is "Nissan".
The Lowest average price of make is "Toyota".


**10.How many cars manufactures(Make) in a year?**

In [ ]:
year = 2014
filtered_df = df[df['Year']==year]
car_makes = filtered_df['Make'].unique()
car_makes

array(['Toyota', 'Ford', 'Chevrolet', 'Honda', 'Nissan'], dtype=object)

from above, in the year 2014, manufacturers makes number of cars in the year 2014 is five (make) which are, 'Toyota', 'Ford', 'Chevrolet', 'Honda', 'Nissan'.

In [ ]:
year==2016
manufacturers_in_year = df[df['Year'] == year]['Make'].nunique()
manufacturers_in_year

5

from above, in the year 2016, manufacturers makes number of cars in the year 2014 is five (make) 5.

**11.How many types of model in a year 2016 & 2019?**

In [ ]:
filtered_2016 = df[df['Year']==2016]
filtered_2019 = df[df['Year']==2019]
car_model_2016 = filtered_2016['Model'].unique()
car_model_2019= filtered_2019['Model'].unique()
car_model_2016
car_model_2019

array(['Civic', 'Altima', 'Silverado', 'Camry', 'F-150'], dtype=object)

from above , types of model in a year 2016 & 2019 are Civic', 'Altima', 'Silverado', 'Camry', 'F-150'.

**12. What are categories of car conditions, mention with their names?**

In [ ]:
car_cond =df['Condition'].unique()
car_cond

array(['Excellent', 'Good', 'Fair'], dtype=object)

The categories of car conditions are 'Excellent', 'Good', 'Fair'.

**13.What types of cars from which manufacturer where mileage is greater than 13000 and price is less than 27800?**

In [ ]:
df[(df['Mileage'] >13000) & (df['Price'] < 27800)]

Make      Model  Year  Mileage  Condition     Price
0         Ford  Silverado  2022    18107  Excellent  19094.75
1       Toyota  Silverado  2014    13578  Excellent  27321.10
2    Chevrolet      Civic  2016    46054       Good  23697.30
3         Ford      Civic  2022    34981  Excellent  18251.05
4    Chevrolet      Civic  2019    63565  Excellent  19821.85
..         ...        ...   ...      ...        ...       ...
994      Honda  Silverado  2015    50874  Excellent  24456.40
995     Nissan      Camry  2010   149032  Excellent  24548.50
996  Chevrolet      F-150  2014    20608  Excellent  26969.70
997       Ford     Altima  2016   109851       Good  20507.55
999     Nissan  Silverado  2017   128390  Excellent  18580.60

[882 rows x 6 columns]

these are the cars from the manufacturers where mileage is greater than 13000 and price is less than 27800.

1.Is there a relationship between the year of the car and its price?
2.How does the condition of the car affect its price?
3.What is the correlation between price and numerical features (mileage, year)?
4.What is the average price of cars for each make and model?
5.Which brands and models have the highest and lowest average prices?
6.How have car prices changed over the years?
7.Is there a noticeable trend in pricing based on the manufacturing year?
8.Is there a correlation between the price of the car and its mileage?
9.Do cars with higher mileage tend to have lower prices?
10.How does the condition of the car (Excellent, Good, Fair) affect its price?
11.Are there significant differences in prices based on the condition of the car?
12.Can you compare the average prices of cars from different manufacturers?
13.How do prices vary between different car models within the same manufacturer?
14.Calculate summary statistics like mean, median, and standard deviation of car prices to describe the central tendency and variability of the data.

**14.How have car prices changed over the years?**

In [ ]:
car_p_changed_yrs=df.groupby('Year')['Price'].mean()
car_p_changed_yrs


Year
2010    27908.863855
2011    27043.655435
2012    25818.688732
2013    25269.742466
2014    24209.373118
2015    23001.527419
2016    22089.416883
2017    21009.512162
2018    19944.909559
2019    19172.257059
2020    18278.516438
2021    16607.056494
2022    16400.130556
Name: Price, dtype: float64

Car prices changes over the years from year 2010 is 27908.86 , then in year 2011 is 27043.65 , then increasing next 2 years then decreasing for further some years and so on..

**15.Is there a noticeable trend in pricing based on the manufacturing year?**

In [ ]:
price_per_change= car_p_changed_yrs.pct_change()*100
price_per_change.reset_index()

Year     Price
0   2010       NaN
1   2011 -3.100121
2   2012 -4.529590
3   2013 -2.126159
4   2014 -4.196202
5   2015 -4.989166
6   2016 -3.965435
7   2017 -4.888788
8   2018 -5.067241
9   2019 -3.873933
10  2020 -4.661635
11  2021 -9.144396
12  2022 -1.246012

In [ ]:
price_per_change= car_p_changed_yrs.pct_change()*100
price_per_change

Year
2010         NaN
2011   -3.100121
2012   -4.529590
2013   -2.126159
2014   -4.196202
2015   -4.989166
2016   -3.965435
2017   -4.888788
2018   -5.067241
2019   -3.873933
2020   -4.661635
2021   -9.144396
2022   -1.246012
Name: Price, dtype: float64

In [ ]:
df1= df[['Price','Year']]
df2=price_per_change


In [ ]:
merged_df = pd.merge(df1, df2, on='Year')
merged_df

Price_x  Year   Price_y
0    19094.75  2022 -1.246012
1    18251.05  2022 -1.246012
2    17968.35  2022 -1.246012
3    18296.00  2022 -1.246012
4    18965.45  2022 -1.246012
..        ...   ...       ...
995  29060.05  2011 -3.100121
996  24141.05  2011 -3.100121
997  26913.00  2011 -3.100121
998  28146.60  2011 -3.100121
999  29126.20  2011 -3.100121

[1000 rows x 3 columns]

there is percentage change in years is fluctuating from year 2011 to 2022.

**16.How does the condition of the car (Excellent, Good, Fair) affect its price?**

In [ ]:
avg_p_by_condition=df.groupby('Condition')['Price'].mean().sort_values(ascending=False)
avg_p_by_condition

Condition
Fair         22357.820796
Good         22222.674315
Excellent    22150.842017
Name: Price, dtype: float64

from above , the condition affects the price of the cars , fair condition getting high price however, good condition getting moderate and excellent condition lower price.

***Are there significant differences in prices based on the condition of the car? ***ASK

**17.Can you compare the average prices of cars from different manufacturers? ***

In [ ]:
avg_p_make= df.groupby('Make')['Price'].mean().sort_values(ascending=True)
avg_p_make

Make
Toyota       21881.913415
Chevrolet    21957.203589
Honda        22238.084171
Ford         22354.415578
Nissan       22587.502128
Name: Price, dtype: float64

average price of the cras from different manufacturers , the higher mean price for make is "Nissan" 22587.50 and the lower mean price for make is "Toyota" "21881.91"

**18.How do prices vary between different car models with the same manufacturer? ***

In [ ]:
avg_p_vary= df.groupby(['Make','Model'])['Price'].mean().sort_values(ascending=True)
avg_p_vary

Make       Model    
Honda      Silverado    21291.349020
Chevrolet  Altima       21407.802000
Nissan     Altima       21493.059783
Honda      Altima       21520.601250
Toyota     Altima       21547.499038
Ford       F-150        21627.950000
           Camry        21660.350000
Chevrolet  Civic        21721.005263
Toyota     Civic        21735.738158
           Camry        21762.431395
Honda      Civic        21779.972973
Chevrolet  Camry        21912.704878
Nissan     Civic        22029.708065
Toyota     Silverado    22105.582857
Ford       Silverado    22250.963889
Chevrolet  Silverado    22396.186765
Toyota     F-150        22429.305405
Nissan     Silverado    22429.386111
Chevrolet  F-150        22464.695652
Ford       Civic        22853.082895
Honda      F-150        22866.732258
Nissan     Camry        23144.928750
Ford       Altima       23576.052632
Nissan     F-150        24045.532857
Honda      Camry        24099.205000
Name: Price, dtype: float64

from above prices vary between different car models with the same manufacturer.. make "Honda" model "silverado" has lower prices 21291.34 however,
"Honda" "Camry" has higher price 24099.20

**19.Calculate summary statistics like mean, median, and standard deviation of car prices to describe the central tendency and variability of the data.**

In [ ]:
mean_price= df['Price'].mean()
mean_price


22195.20565

mean price of price which is 22195.20

In [ ]:
median_price= df['Price'].median()
median_price



22247.875

median price of price is 22247.875.

In [ ]:
sd_price = df['Price'].std()
sd_price

4245.191585398393

standard deviation or variation of price is 4245.1915

**20.Calculate summary statistics like mean, median, and standard deviation of car mileage to describe the central tendency and variability of the data.**

In [ ]:
mean_mile=df["Mileage"].mean()
mean_mile


78796.927

mean mileage of mileage is 78796.927

In [ ]:
median_mile=df["Mileage"].median()
median_mile


78056.5

median mileage of mileage is 78056.5

In [ ]:
sd_mile=df["Mileage"].std()
sd_mile


39842.25994085759

variation in milaege or sd of milaege is 39842.2599

In [ ]:
df.columns


Index(['Make', 'Model', 'Year', 'Mileage', 'Condition', 'Price'], dtype='object')

In [ ]:
corr= df['Mileage'].corr(df['Price'])
corr

-0.45425031332318344

In [ ]:
corr= df['Year'].corr(df['Price'])
corr

-0.8830933715991623

In [ ]:
df.head(2)

Make      Model  Year  Mileage  Condition     Price
0    Ford  Silverado  2022    18107  Excellent  19094.75
1  Toyota  Silverado  2014    13578  Excellent  27321.10

In [ ]:
#from sklearn.preprocessing import LabelEncoder

#label_encoder = LabelEncoder()

In [ ]:
#encoded_Model= label_encoder.fit_transform(Model)

In [ ]:
#mean_car_price_by_fuel_type = df.groupby('Fuel_Type')['Car_Price'].mean()
#df['Fuel_Type_Encoded'] = df['Fuel_Type'].map(mean_car_price_by_fuel_type)
